datashader/transfer_functions/__init__.py

```diff
@@ -574,20 +577,102 @@ def spread(img, px=1, shape='circle', how='over', mask=None, name=None):
         raise ValueError("mask must be a square 2 dimensional ndarray with "
                          "odd dimensions.")
         mask = mask if mask.dtype == 'bool' else mask.astype('bool')
-    kernel = _build_spread_kernel(how)
+    if how is None:
+        how = 'over' if is_image else 'add'
+
     w = mask.shape[0]
     extra = w // 2
-    M, N = img.shape
-    buf = np.zeros((M + 2*extra, N + 2*extra), dtype='uint32')
-    kernel(img.data, mask, buf)
-    out = buf[extra:-extra, extra:-extra].copy()
-    return Image(out, dims=img.dims, coords=img.coords, name=name)
+    M, N = img.shape[:2]
+    padded_shape = (M + 2*extra, N + 2*extra)
+    float_type = img.dtype in [np.float32, np.float64]
+
+    if is_image:
+        kernel = _build_spread_kernel(how, is_image)
+    elif float_type:
+        kernel = _build_float_kernel(how, w)
+    else:
+        kernel = _build_int_kernel(how, w, img.dtype == np.uint32)
+
+    if is_image:
+        buf = np.zeros(padded_shape, dtype=img.dtype)
+        kernel(img.data, mask, buf)
+        out = buf[extra:-extra, extra:-extra].copy()
+    elif float_type:
+        buf = np.full(padded_shape, np.nan, dtype=img.dtype)
+        kernel(img.data, mask, buf)
+        out = buf[extra:-extra, extra:-extra].copy()
+    elif len(img.shape)==2:
+        buf = np.zeros(padded_shape, dtype=img.dtype)
+        kernel(img.data, mask, buf)
+        out = buf[extra:-extra, extra:-extra].copy()
+    else:
+        layers = []
+        for category in range(img.shape[2]):
+            layer = img[:,:,category]
+            if float_type:
+                buf = np.full(padded_shape, np.nan, dtype=img.dtype)
+            else:
+                buf = np.zeros(padded_shape, dtype=layer.dtype)
+            kernel(layer.data, mask, buf)
+            layers.append(buf[extra:-extra, extra:-extra].copy())
+            out = np.dstack(layers)
```

Comment:
```suggestion
        out = np.dstack(layers)
```

Seems like it would have worked, but with the previous indentation it would have overwritten `out` every time a layer is processed, which probably was not intended?)

Reply:
This is totally untested pseudocode, but I think this whole section of the code would be much more readable and maintainable if it's boiled down to what it actually does:

```
    fill_value = np.nan if float_type else 0

    def apply_kernel(layer):
        buf = np.full(padded_shape, fill_value, dtype=layer.dtype)
        kernel(layer.data, mask, buf)
        return buf[extra:-extra, extra:-extra].copy()
        
    if len(img.shape)==2:
        out = apply_kernel(img)
    else:
        out = np.dstack([apply_kernel(img[:,:,category])
                        for category in range(img.shape[2])]
```

Otherwise we can't see that what's happening in all these cases is quite similar, with only trivial differences that are hard to spot otherwise.